### DISCUSSION TITLE
Proposal: rockspec v4 specs

### DISCUSSION THREAD

**Discussion by NTBBloodbath** [keplerproject/luarocks#1729] [Ideas]
Hi, hope you're doing well!

I've been working with luarocks for a while now and I must say that my overall experience has been very pleasant, thanks for taking on the project! However, I have found some discrepancies and other things I would like to see in the rockspec format.

### Separate concerns

Currently, the `description` table mixes general metadata (like `license`, `homepage` and `maintainer`) with the actual description of the package. This can be confusing, and thus I propose either separating this information into a dedicated `metadata` table or renaming the existing table to `metadata`.

> Example of separate tables:
```lua
metadata = {
  license = "MIT",
  homepage = "https://example.com",
  maintainer = "Your Name <your.email@example.com>"
}

description = {
  summary = "A short summary of the package.",
  detailed = "A more detailed description of the package."
}
```

> Example of renamed table:
```lua
metadata = { -- Renamed from 'description'
  license = "MIT",
  homepage = "https://example.com",
  maintainer = "Your Name <your.email@example.com>",
  description = { -- Moved here
	summary = "A short summary of the package."
	detailed = "A more detailed description of the package."
  }
}
```

### Add additional metadata fields

Adding fields for `README` and `CHANGELOG` paths would be extremely helpful for automated tooling and documentation generation. This could also benefit the luarocks site, allowing it to display more information about the rocks, for example a README file instead of a simple description.

### Add development dependencies

After a very long time programming, I miss dependencies that are exclusively for the purpose of being used during development. It's crucial to distinguish between runtime dependencies and dependencies needed only for development (linting, building, etc.), for example I don't want users of my libraries to have to install tools like luacheck if it's not really necessary for the library to work. Something like an extra table called `dev_dependencies` which follows exactly the same syntax as `dependencies` could be ideal.

```lua
dependencies = {
  "lua >= 5.1",
  "lpeg >= 1.0"
}

dev_dependencies = {
  "luacheck >= 0.23" -- Example linter
}
```

### Add opt-in dependencies support

This one is pretty much inspired by languages like Rust, where its package manager `cargo` has "`features`" which enables certain capabilities in the libraries/binaries and thus decides which dependencies to download based on the specific functionalities desired by the users installing them. This could significantly reduce the overall size of the installed rocks as well.

Here is a dummy example of what I'm looking for:
```lua
dependencies = {
  "lua >= 5.1"
}

features = {
  ["json"] = { -- Feature name
    description = "Support for JSON encoding/decoding",
    dependencies = {
      "dkjson >= 2.5"
    }
  },
  ["xml"] = { -- Feature name
    description = "Support for XML parsing",
    dependencies = {
      "lua-expat >= 1.3"
    }
  }
}
```

Users could then install the rock with a specific feature using a command like `luarocks install --with-features=json my-package`. If multiple features are desired, a comma could be used to separate them and make it easy to parse the command flag option.

---

I believe these changes would greatly improve the rockspec format and the overall LuaRocks ecosystem. I'm open to feedback and further discussion.

Cheers!

**Comment by mrcjkb** [comment]
To add to this: It would be nice if luarocks supported something like this:

```lua
git_dependencies = {
  "github:<owner>/<repo>"
  "gitlab:<owner>/<repo>?ref=<branch-or-tag-name>&rev=<commit-sha>"
}
```

If there is a n scm-rockspec in the project root, that could be used to build. Otherwise, default to the `none` build type.

This would allow luarocks projects to depend on lua libraries without uploading a rockspec to the root manifest (which some authors may be reluctant to do).

**Reply by hishamhm** [reply]
adding a pseudo-protocol for particular websites doesn't sound ideal as that favors some Git hosting options over others, but perhaps `git:` with the full URL could work?

**Reply by mrcjkb** [reply]
I based my suggestion on [the nix flakes input schema](https://nix.dev/manual/nix/2.24/command-ref/new-cli/nix3-flake.html#flake-inputs), for which we have implemented something similar in [rocks-git.nvim](https://github.com/nvim-neorocks/rocks-git.nvim/).

It supports pseudo URLs with `github:`, `gitlab:`, `codeberg:` and `sourcehut:` prefixes, and also full git HTTPS or SSH URLs.

Of course, just a full git URL is simpler implementation-wise and you don't gain anything but syntactic sugar from the pseudo-URLs. So I don't feel strongly about them.

**Comment by hishamhm** [comment]
> Users could then install the rock with a specific feature using a command like luarocks install --with-features=json my-package. If multiple features are desired, a comma could be used to separate them and make it easy to parse the command flag option.

`feature` support is really cool, but it does add a ton of complexity, in that you have to keep track of which features your existing packages were built with, and what features packages depend on. So now, if you have installed `foo` without feature `X`, and then install `bar` which depends on feature `Y`, you need to reinstall `foo`. But then features `X` and `Y` may be incompatible with each other, so installing one would break another dependency `baz` which depends on the other feature. There are multiple ways to solve this (e.g. the "install everything multiple times" like npm, "support multiple variants of a package but use a smart loader" like LuaRocks tries to do), but each has its pros and cons. I'm not saying I'm against the idea, just saying that even getting to a fleshed-out proposal on how exactly such a feature should work is substantial work (covering how to store/fetch the metadata, how to deal with conflicts, etc.)
